### DISCUSSION TITLE
mpremote on a STM32WL5x board, can't load files

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#13157] [Using MicroPython] [answered]
Because I couldn’t talk to a STM32WL5x board using rshell I started using mpremote.  This seems to work most of the time, except when trying to copy files bigger than say 500-1000 bytes.  The only observation is that every time I try the transfer it will get to a different percentage, ie 4% - 75%.  Often it will never transfer anything.  This distro is using LFS2 and I appear to have 30-45K of flash to play with.

I have been using mpremote on a STM32F411 board without this problem.

Do you have to do something special to use mpremote on a NUCLEO_WL55 board?

**Comment by davefes** [comment]
Found a clue here:
https://github.com/micropython/micropython/pull/8495

`mpremote cp :large_file . > output`

Maybe, this should go into the docs?

**Comment by andrewleech** [comment, marked as answer]
This is a problem on all ports / boards which use UART for repl and internal flash for storage. With internal flash, when code is running from the same flash  the CPU is paused while flash write / erase operations take place. These pauses can cause UART bytes to be dropped.

I've got a fix for this on STM32 in https://github.com/micropython/micropython/pull/9054 (which has links to earlier discussions and investigations) but it's a rather involved change.

Note the F411 doesn't have this issue because it's using USB rather than UART, the USB has flow control so pauses the communication while flash operations take place.
UART with RTS/CTS flow control also fixes the problem, but most stm boards like all the nucleo ones don't wire these lines up.

**Reply by davefes** [reply]
Yes, I got "lost" near the beginning of #9054.
Thanks for pointing out the difference between the two platforms.
I'll see how the "work-around" holds-out.

**Reply by davefes** [reply]
The "clue" above works for transferring files from the remote device to the host.  It doesn't seem to work the other way around.

**Reply by davefes** [reply]
Unfortunately, the E77-400M22S uses LP_UART1_CTS (PA6) for RX/TX switching.  LP_UART1_RTS (PA1) is free to use :(  When could this fix be merged?

**Reply by andrewleech** [reply]
Yeah transferring from device to PC is always fine; reading the flash doesn't cause CPU pause, only writing.

**Reply by davefes** [reply]
@andrewleech ,  Just a "left-field" suggestion ... could repl use USART1 or 2?  This would give RTS/CTS for flow control.  PA0 (the user button) might need moving.   This would free-up PA6.

What would be involved in getting flow-control?
https://forum.micropython.org/viewtopic.php?t=1669
Seems at the device end, maybe not an issue.  But how would you handle it at the host-end running say mpremote?

**Reply by andrewleech** [reply]
You can use a different UART certainly, and enabling flow control is straightforward. It needs the extra couple of pins wired to the USB UART chip, sounds like you already know that.
On the PC side it's just an extra function call in pyserial to enable flow control. I don't know if mpremote has support for that already or not, but wouldn't be hard to add.

I don't have a timeframe for my other pr fixing this to get merged, I expect a fair bit of maintainer testing would be needed.

**Reply by davefes** [reply]
@andrewleech  Thanks for the added info, I will investigate further.

**Reply by davefes** [reply]
@andrewleech Having another think about this issue.  I am able to load Micropython on using STM32 CubeProgrammer.  Another a few hours searching it is not clear if one can upload `.py` files to flash.  If one can then I would need to know where to flash them to.

It might not be a nice workflow, but at least I might be able to get code running on the  E77-400M22S.

Comments?

**Reply by andrewleech** [reply]
You can freeze code into the firmware so it's already included when you flash the firmware.
https://docs.micropython.org/en/latest/reference/manifest.html

It is possible to create a snapshot filesystem too with your chosen files and then flash that on in the right location where the standard filesystem goes.

I've done this with the Unix port of micropython to create a dfu file to flash onto  stm alongside the main firmware, here's an example: https://gist.github.com/andrewleech/3264bc910afdaf5809be99aa673766a8

**Reply by davefes** [reply]
I froze one of my first ESP32 projects, however I find that code changes so many times that it was easier to just get a bigger flash.
I will re-consider that approach and thanks for the example `create a dfu file`.
